### PR TITLE
Bloom filter resizing

### PR DESCRIPTION
**Pull request by ejfitzgerald**

### CODE REVIEW

File: libs/bloom-filter/tools/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,3 @@
+cmake_minimum_required(VERSION 3.10 FATAL_ERROR)
```
Comment (ejfitzgerald) [inline libs/bloom-filter/tools/CMakeLists.txt:4]: Beautify

File: libs/bloom-filter/tools/main.cpp [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+  }
+
+  return output;
+}
+
+int main(int argc, char const *const *argv)
+{
+  if (argc != 3)
+  {
+    std::cerr << "Usage: bloom-tool <filter size> <hashes file>" << std::endl;
```
Comment (ejfitzgerald) [inline libs/bloom-filter/tools/main.cpp]: bit size

File: libs/bloom-filter/src/historical_bloom_filter.cpp [hunk 3]
```diff
@@ -90,33 +93,20 @@ struct BinarySerializer<bloom::HistoricalBloomFilter::CacheEntry, D>
...
-  key.Resize(32);
-  assert(key.size() == 32);
-
-  // write the bucket as the value into the key
-  *reinterpret_cast<uint64_t *>(key.pointer()) = bucket;
-
-  return storage::ResourceID{key};
-}
+constexpr std::size_t BLOOM_FILTER_SIZE = 160 * 8 * 1024 * 1024;
+constexpr uint64_t STORAGE_SECTOR_SIZE  = BLOOM_FILTER_SIZE + (BLOOM_FILTER_SIZE >> 3);  // 112.5 %
```
Comment (ejfitzgerald) [inline libs/bloom-filter/src/historical_bloom_filter.cpp]: "hands waived"

File: libs/bloom-filter/src/historical_bloom_filter.cpp [hunk 4]
```diff
@@ -303,7 +328,13 @@ std::size_t HistoricalBloomFilter::TrimCache()
       // flush the page if dirty
       if (it->second.dirty)
       {
-        SaveBucketToStore(it->first, it->second);
+        if (!SaveBucketToStore(it->first, it->second))
+        {
+          // can't flush the page to disk, for saftey keep the page in memory
+          FETCH_LOG_ERROR(LOGGING_NAME, "Unable to flush bucket: ", it->first, " to disk");
```
Comment (ejfitzgerald) [inline libs/bloom-filter/src/historical_bloom_filter.cpp:339]: telemetry

File: libs/bloom-filter/src/historical_bloom_filter.cpp [hunk 5]
```diff
@@ -441,36 +468,64 @@ bool HistoricalBloomFilter::LookupBucketFromStore(uint64_t bucket, CacheEntry &e
...
+    serializers::LargeObjectSerializeHelper serialiser{};
+    serialiser << entry;
 
-  // store the cache on disk
-  store_.Set(key, entry);
+    bloom_filter_buffer = serialiser.data();
+  }
+  catch (std::exception const &ex)
+  {
+    FETCH_LOG_ERROR(LOGGING_NAME, "Error serialising bloom filter entry: ", ex.what());
```
Comment (ejfitzgerald) [inline libs/bloom-filter/src/historical_bloom_filter.cpp:524]: return false

File: libs/storage/src/fixed_size_journal.cpp [hunk 6]
```diff
@@ -0,0 +1,521 @@
...
+ * @param sector The sector index to read
+ * @param buffer The buffer to populate with the contents
+ * @return true if successful, otherwise false
+ */
+bool FixedSizeJournalFile::Get(uint64_t sector, byte_array::ConstByteArray &buffer) const
+{
+  FETCH_LOCK(lock_);
+
+  // advance to the sector and read the entire sector
+#if 1
```
Comment (ejfitzgerald) [inline libs/storage/src/fixed_size_journal.cpp]: Remove!

File: libs/storage/src/fixed_size_journal.cpp [hunk 7]
```diff
@@ -0,0 +1,521 @@
...
+ * @return true if successful, otherwise false
+ */
+bool FixedSizeJournalFile::Set(uint64_t sector, byte_array::ConstByteArray const &buffer)
+{
+  FETCH_LOCK(lock_);
+
+  // ensure that we can save this much data
+  if (buffer.size() > max_sector_data_size_)
+  {
+    return false;
```
Comment (ejfitzgerald) [inline libs/storage/src/fixed_size_journal.cpp:409]: Add log message here

File: libs/storage/src/fixed_size_journal.cpp [hunk 8]
```diff
@@ -0,0 +1,521 @@
...
+ * @return true if successful, otherwise false
+ */
+bool FixedSizeJournalFile::Flush()
+{
+  FETCH_LOCK(lock_);
+
+  return InternalFlush();
+}
+
+bool FixedSizeJournalFile::Clear(std::string const &filename)
```
Comment (ejfitzgerald) [inline libs/storage/src/fixed_size_journal.cpp:461]: Add docs and mention thread satey

File: libs/storage/src/fixed_size_journal.cpp [hunk 9]
```diff
@@ -0,0 +1,521 @@
...
+ * @return true if successful, otherwise false
+ */
+bool WriteSectorHeader(std::fstream &stream, uint64_t data_size)
+{
+  // populate the buffer
+  std::array<uint8_t, SECTOR_HEADER_SIZE> buffer{};
+
+  WriteBigEndian(&buffer[0], SYNC_BYTE);
+  WriteBigEndian(&buffer[1], data_size);
+
```
Comment (ejfitzgerald) [inline libs/storage/src/fixed_size_journal.cpp:228]: static assert

File: libs/storage/src/fixed_size_journal.cpp [hunk 10]
```diff
@@ -0,0 +1,521 @@
...
+bool WriteFileHeader(std::fstream &stream, FileHeader const &header)
+{
+  // populate the header buffer
+  std::array<uint8_t, FILE_HEADER_SIZE> buffer{};
+  WriteBigEndian(&buffer[0], FILE_MAGIC);           // (8)
+  WriteBigEndian(&buffer[8], header.version);       // (2)
+  WriteBigEndian(&buffer[10], header.sector_size);  // (8)
+  WriteBigEndian(&buffer[18], header.num_sectors);  // (8)
+
```
Comment (ejfitzgerald) [inline libs/storage/src/fixed_size_journal.cpp:172]: static assert here too

**Review by n-hutton — DISMISSED** [review]

File: libs/storage/src/fixed_size_journal.cpp [hunk 11]
```diff
@@ -0,0 +1,533 @@
...
+  // check to see if the open was successful
+  bool const success = stream_.is_open() && !(stream_.bad() || stream_.fail());
+  if (!success)
+  {
+    stream_.clear();  // reset stream flags
+    FETCH_LOG_ERROR(LOGGING_NAME, "Failed to open requested file");
+    return false;
+  }
+
+  // perform a flush, this will make sure that the
```
Comment (wkaluza) [inline libs/storage/src/fixed_size_journal.cpp]: Comment not fully flushed. :grin:

**Review by wkaluza — DISMISSED** [review]
Looks good to me, see minor comment above.

**Review by wkaluza — DISMISSED** [review]

**Review by n-hutton — DISMISSED** [review]

**Pull request merged**